File: DP/139. Word Break.py
"""
Freewheel 当年的面试题，也是一道经典题。

"""
class Solution:
    def wordBreak(self, s: str, words: List[str]) -> bool:
        words = set(words)
        if not s or s in words:
            return True
        dp = [False] * (len(s) + 1)
        dp[0] = True
        for i in range(len(s)):
            for j in range(i, len(s) + 1):
                if dp[i] and s[i:j] in words:
                    dp[j] = True
        return dp[-1]

File: DP/139 wordbreak.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        set<string> st;
        vector<bool> v(s.size() + 1, false);
        if(s.size() && wordDict.size())
            v[0] = true;
        else
            return false;
        for(auto word:wordDict)
            st.insert(word);
        for(int i = 0; i < s.size(); i++)
        {
            for(int j = i + 1; j < s.size() + 1; j++)
            {
                if(v[i] && st.count(s.substr(i, j - i)))
                {
                    v[j] = true;
                }
            }
        }
        return v[s.size()];
    }
};
